### PR TITLE
Avoid `.packages` call when packages pane disabled

### PR DESCRIPTION
**Pull request by kevinushey**
Not quite ready for merge -- just putting it up there for now...

## TODO

- [x] Avoid calling `updatePackageEvents()` when packages pane disabled
- [x] Rip out `onPackageLoaded` handler in `SessionShiny`
- [ ] Provide a new event, `onProjectPackageLoaded`, for use in `BreakpointManager`
- [ ] Verify no other code paths use this (check `onPackageStatusChanged`, `onPackageLoaded`, `onPackageUnloaded` -- e.g. `ag -Q " onPackage"`)

### CODE REVIEW

File: src/cpp/session/modules/SessionPackages.cpp [hunk 1]
```diff
@@ -537,44 +546,22 @@ Error initialize()
...
+   RS_REGISTER_CALL_METHOD(rs_packagesPaneEnabled, 0);
+
    using boost::bind;
    using namespace module_context;
    ExecBlock initBlock ;
    initBlock.addFunctions()
       (bind(sourceModuleRFile, "SessionPackages.R"))
       (bind(registerRpcMethod, "available_packages", availablePackages))
-      (bind(registerRpcMethod, "get_package_state", getPackageState))
-      (bind(r::exec::executeString, ".rs.packages.initialize()"));
```
Comment (kevinushey) [inline src/cpp/session/modules/SessionPackages.cpp]: @jjallaire, I moved this to deferred init because otherwise this code tries to run before the native routines have actually been registered (and so we can't successfully check `.rs.routines$rs_packagesPaneEnabled`).

Do you know if there's any code that depends on `.rs.packages.initialize()` running eagerly or do you think a deferred init should be okay? (Since it just adds a bunch of hooks / handlers it seems like a deferred init should be okay)

**Comment by jjallaire** [comment]
This has a bit more refactoring than I'd typically hope to see in a change that needs to be backported. That said, all of your changes are good ones and Uwe seems okay with the state of things for now so making the "right" rather than merely "safe" seems like goodness. When we are ready to merge we can still rebase into a single commit and then consider backporting this after we have more confidence based on usage of the daily build.

**Comment by kevinushey** [comment]
One other thought that would definitely speed things up:

Rather than calling `.packages(TRUE)`, which verifies that each directory within a particular library path actually is the path to an R package (ie -- implies many `file.exists()` calls), we could just indiscriminately take all directory names within the library folder, and add hooks to those names.

Rationale:

- Would definitely capture all packages as before
- Things that aren't actually packages would have hooks registered, but those hooks would never actually be fired (since no such package action with that name would occur) and so should not cause issues

In other words, we'd replace `.packages(TRUE)` with something like `list.dirs(.libPaths(), recursive = FALSE, full.names = FALSE)` and assume that the directory names returned are indeed package names.

**Comment by kevinushey** [comment]
There we go! A much tidier PR. :)

File: src/cpp/session/modules/SessionPackages.R [hunk 2]
```diff
@@ -62,7 +62,12 @@ if (identical(as.character(Sys.info()["sysname"]), "Darwin") &&
       .Call(.rs.routines$rs_packageUnloaded, pkgname)
    }
    
-   sapply(.packages(TRUE), function(packageName) 
+   pkgNames <- if (exists("list.dirs", envir = .BaseNamespaceEnv))
```
Comment (jjallaire) [inline src/cpp/session/modules/SessionPackages.R]: Would this be more clear as a check for R >= 2.13

**Comment by kevinushey** [comment]
Okay, the check is now keyed off of `getRversion() >= "3.0.0"`.
